pkg/globalnet/controllers/ipam/globalnet_ingress.go

```diff
@@ -0,0 +1,82 @@
+package ipam
+
+import (
+	"crypto/sha256"
+	"encoding/base32"
+	"fmt"
+	"strings"
+
+	"github.com/coreos/go-iptables/iptables"
+	"github.com/submariner-io/submariner/pkg/util"
+	k8sv1 "k8s.io/api/core/v1"
+	"k8s.io/klog"
+)
+
+func (i *Controller) createNecessaryIPTableChains() error {
+	ipt, err := iptables.New()
+	if err != nil {
+		return fmt.Errorf("error initializing iptables: %v", err)
+	}
+
+	klog.V(4).Infof("Install/ensure %s chain exists", submarinerGlobalNet)
+	if err = util.CreateChainIfNotExists(ipt, "nat", submarinerGlobalNet); err != nil {
+		return fmt.Errorf("error creating iptables chain %s: %v", submarinerGlobalNet, err)
+	}
+
+	forwardToSubGlobalNetRuleSpec := []string{"-j", submarinerGlobalNet}
+	if err = util.PrependUnique(ipt, "nat", "PREROUTING", forwardToSubGlobalNetRuleSpec); err != nil {
+		klog.Errorf("error appending iptables rule \"%s\": %v\n", strings.Join(forwardToSubGlobalNetRuleSpec, " "), err)
+	}
+	return nil
+}
+
+func (i *Controller) updateIngressRulesForService(globalIP, chainName string, addOrDelete Operation) error {
+	ipt, err := iptables.New()
+	if err != nil {
+		return fmt.Errorf("error initializing iptables: %v", err)
+	}
+
+	ruleSpec := []string{"-d", globalIP, "-j", chainName}
+	if addOrDelete == AddRules {
+		klog.V(4).Infof("Installing iptables rule for Service %s", strings.Join(ruleSpec, " "))
+		if err = ipt.AppendUnique("nat", submarinerGlobalNet, ruleSpec...); err != nil {
+			return fmt.Errorf("error appending iptables rule \"%s\": %v\n", strings.Join(ruleSpec, " "), err)
+		}
+	} else if addOrDelete == DeleteRules {
+		klog.V(4).Infof("Deleting iptable ingress rule for Service: %s", strings.Join(ruleSpec, " "))
+		if err = ipt.Delete("nat", submarinerGlobalNet, ruleSpec...); err != nil {
+			return fmt.Errorf("error deleting iptables rule \"%s\": %v\n", strings.Join(ruleSpec, " "), err)
+		}
+	}
+	return nil
+}
+
+func (i *Controller) kubeProxyClusterIpServiceChainName(service *k8sv1.Service) string {
+	// CNIs that use kube-proxy with iptables for loadbalancing create an iptables chain for each service
+	// and incoming traffic to the clusterIP Service is directed into the respective chain.
+	serviceName := service.GetNamespace() + "/" + service.GetName() + ":"
+	protocol := strings.ToLower(string(service.Spec.Ports[0].Protocol))
+	hash := sha256.Sum256([]byte(serviceName + protocol))
+	encoded := base32.StdEncoding.EncodeToString(hash[:])
+	return kubeProxyServiceChainPrefix + encoded[:16]
+}
+
+func (i *Controller) doesIptablesChainExist(table, chain string) error {
```

Comment:
Why not use bool and check that? Chain not existing is different error than initializing tables.

Reply:
Why not boolean and error? As of now we're not really checking the error reason right? With current code I think it doesn't matter.

Reply:
> Why not boolean and error? As of now we're not really checking the error reason right? With current code I think it doesn't matter.

I agree.  I think a function that starts with _does_ implies a boolean answer. Non-existence isn't an error condition in this case plus makes it difficult for callers to distinguish non-existence and an actual error.

Reply:
Done